core/request.go

```diff
@@ -72,46 +71,80 @@ func (f RequestFunc) Process(t *TransferRequest) error {
 
 // filleTransferRequest creates HTTP request to transfer a given file name
 // https://matt.aimonetti.net/posts/2013/07/01/golang-multipart-file-upload-example/
-func fileTransferRequest(c CatalogEntry, tr *TransferRequest) (*http.Request, error) {
+func fileTransferRequest(c CatalogEntry, tr *TransferRequest) (*http.Response, error) {
 	file, err := os.Open(c.Pfn)
+	defer file.Close()
 	if err != nil {
 		return nil, err
 	}
-	defer file.Close()
-	body := &bytes.Buffer{}
-	writer := multipart.NewWriter(body)
+
+	pipeOut, pipeIn := io.Pipe()
+	writer := multipart.NewWriter(pipeIn)
+	// do the request concurrently
+	var resp *http.Response
+	done := make(chan error)
+	go func() {
+		// prepare request
+		url := fmt.Sprintf("%s/upload", tr.DstUrl)
+		req, err := http.NewRequest("POST", url, pipeOut)
+		if err != nil {
+			done <- err
+			return
+		}
+
+		req.Header.Set("Content-Type", writer.FormDataContentType())
+		req.Header.Set("Pfn", c.Pfn)
+		req.Header.Set("Lfn", c.Lfn)
+		req.Header.Set("Bytes", fmt.Sprintf("%d", c.Bytes))
+		req.Header.Set("Hash", c.Hash)
+		req.Header.Set("Src", tr.SrcAlias)
+		req.Header.Set("Dst", tr.DstAlias)
+		req.Header.Set("Content-Type", writer.FormDataContentType())
+		client := utils.HttpClient()
+		resp, err = client.Do(req)
+		if err != nil {
+			done <- err
+			return
+		}
+
+		done <- nil
+	}()
+
 	part, err := writer.CreateFormFile("data", filepath.Base(c.Pfn))
 	if err != nil {
 		return nil, err
 	}
+
 	_, err = io.Copy(part, file)
```

Comment:
@vkuznet here it is copying the files in the chunk. `io.copy` copies the data and put it in a pipe. It will not copy the data further until and unless it has space in the pipe.